core/services/ocr2/plugins/ocr2keeper/evm21/registry.go

```diff
@@ -277,27 +280,93 @@ func (r *EvmRegistry) refreshActiveUpkeeps() error {
 //
 // TODO: check for updated config for log trigger upkeeps and update it, currently we ignore them.
 func (r *EvmRegistry) refreshLogTriggerUpkeeps(ids []*big.Int) error {
-	logTriggerIDs := make([]*big.Int, 0)
+	var err error
+	for i := 0; i < len(ids); i += logTriggerRefreshBatchSize {
+		end := i + logTriggerRefreshBatchSize
+		if end > len(ids) {
+			end = len(ids)
+		}
+		batch := ids[i:end]
+
+		if batchErr := r.refreshLogTriggerUpkeepsBatch(batch); batchErr != nil {
+			multierr.AppendInto(&err, batchErr)
+		}
+
+		time.Sleep(500 * time.Millisecond)
+	}
+
+	return err
+}
+
+func (r *EvmRegistry) refreshLogTriggerUpkeepsBatch(ids []*big.Int) error {
+	var logTriggerIDs []*big.Int
+	var logTriggerHashes []common.Hash
 	for _, id := range ids {
 		uid := &ocr2keepers.UpkeepIdentifier{}
 		if ok := uid.FromBigInt(id); !ok {
 			r.lggr.Warnf("failed to parse upkeep id %s", id.String())
+			continue
 		}
 		switch core.GetUpkeepType(*uid) {
 		case ocr2keepers.LogTrigger:
 			logTriggerIDs = append(logTriggerIDs, id)
-		default:
+			logTriggerHashes = append(logTriggerHashes, common.BigToHash(id))
 		}
 	}
 	newUpkeeps, err := r.logEventProvider.RefreshActiveUpkeeps(logTriggerIDs...)
 	if err != nil {
 		return fmt.Errorf("failed to refresh active upkeep ids in log event provider: %w", err)
 	}
+
+	unpausedLogs, err := r.poller.IndexedLogs(iregistry21.IKeeperRegistryMasterUpkeepUnpaused{}.Topic(), r.addr, 1, logTriggerHashes, indexedLogsConfirmations, pg.WithParentCtx(r.ctx))
+	if err != nil {
+		return err
+	}
+	configSetLogs, err := r.poller.IndexedLogs(iregistry21.IKeeperRegistryMasterUpkeepTriggerConfigSet{}.Topic(), r.addr, 1, logTriggerHashes, indexedLogsConfirmations, pg.WithParentCtx(r.ctx))
+	if err != nil {
+		return err
+	}
+
+	logs := append(unpausedLogs, configSetLogs...)
+
+	configSetBlockNumbers := map[string]uint64{}
+	unpausedBlockNumbers := map[string]uint64{}
+	perUpkeepConfig := map[string][]byte{}
+
+	for _, log := range logs {
+		rawLog := log.ToGethLog()
+		abilog, err := r.registry.ParseLog(rawLog)
+		if err != nil {
+			return err
+		}
+		switch l := abilog.(type) {
+		case *iregistry21.IKeeperRegistryMasterUpkeepTriggerConfigSet:
+			configSetBlockNumbers[l.Id.String()] = rawLog.BlockNumber
+			perUpkeepConfig[l.Id.String()] = l.TriggerConfig
+		case *iregistry21.IKeeperRegistryMasterUpkeepUnpaused:
+			unpausedBlockNumbers[l.Id.String()] = rawLog.BlockNumber
```

Comment:
it seems we can get multiple logs here for the same upkeep? We should only take max values?